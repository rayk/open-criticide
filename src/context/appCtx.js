const assert = require('assert-plus');
const log = require('electron-log');
const R = require('ramda');

const appDetailsReMap = {
  sourceKeys: ['appName', 'updateFeedUrl'],
  destinationKeys: ['name', 'updateUrl', 'jsonStoreRootKey'],
};

const pathMap = {
  sourceKeys: [
    'pathToAppCfg',
    'pathToAppData',
    'pathToAppDir',
    'pathToDocDir',
    'pathToDwnDir',
    'pathToResources',
    'pathToTempDir',
    'pathToUsrHome',
  ],
  destinationKeys: [
    'appFiles',
    'configData',
    'dataFiles',
    'documents',
    'downloads',
    'home',
    'resources',
    'tempData',
  ],
};

const platformMap = {
  sourceKeys: [
    'builtForMacStore',
    'builtForWinStore',
    'chromeVersion',
    'electronVersion',
    'envPlatform',
    'envProcessorArch',
    'executingInDevPath',
  ],
  destinationKeys: [
    'macStoreBuild',
    'winStoreBuild',
    'chromeVersion',
    'electronVersion',
    'os',
    'processor',
    'devMode',
  ],
};

const extractValues = (propertyMap, envSource) =>
  R.zipObj(
    propertyMap.destinationKeys,
    R.props(propertyMap.sourceKeys, envSource),
  );

const prodLogger = () => {
  process.NODE_NDEBUG = false;
  log.transports.file.appName = `${appDetailsReMap.appName}_log`;
  log.transports.file.maxSize = 5 * 1024 * 1024;
  return log;
};

const devLogger = () => {
  process.NODE_NDEBUG = true;
  log.transports.file.file = `logs/${new Date().getTime()}.txt`;
  log.transports.file.level = 'silly';
  log.transports.console.level = 'silly';
  return log;
};

/**
 * Returns a cached Application Context.
 *
 * The context holds the details about the current runtime environment and App configuration.
 * It is cached/memorized after it is created for the first time.
 *
 * @param {object} env - runtime environmental variables.
 * @param {object} config - application configuration.
 * @return {Object} containing the application context.
 */
const appCtx = R.memoize((env, config) => {
  assert.object(config, 'package');
  assert.object(env, 'environment');
  assert(!R.contains(undefined, R.values(env)), 'Not all env props defined.');
  return {
    app: R.merge(extractValues(appDetailsReMap, env), config),
    path: extractValues(pathMap, env),
    platform: extractValues(platformMap, env),
    logger: env.executingInDevPath ? devLogger() : prodLogger(),
    isRunningOnMac: env.envPlatform === 'darwin',
    isRunningOnWin: env.envPlatform === 'win32',
    isRunningOnLinux: env.envPlatform === 'linux',
  };
});

export default appCtx;
