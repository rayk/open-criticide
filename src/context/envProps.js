import { app, autoUpdater } from 'electron';

/**
 * Current runtime Environmental Properties which are available after initialisation.
 * @type {{appName: string, appVersion: string, builtForMacStore: boolean, builtForWinStore: boolean, chromeVersion: string, electronVersion: string, envPlatform: String, envProcessorArch: String, executingInDevPath: boolean, pathToAppCfg: string, pathToAppData: string, pathToAppDir: string, pathToDocDir: string, pathToDwnDir: string, pathToResources: string, pathToTempDir: string, pathToUsrHome: string, updateFeedUrl: string, userLocale: string}}
 */
const envProps = {
  appName: app.getName(),
  appVersion: app.getVersion(),
  builtForMacStore: Boolean(process.mas),
  builtForWinStore: Boolean(process.windowsStore),
  chromeVersion: process.versions.chrome,
  electronVersion: process.versions.electron,
  envPlatform: process.platform,
  envProcessorArch: process.arch,
  executingInDevPath: Boolean(process.execPath.match(/[\\/]electron/)),
  pathToAppCfg: app.getPath('userData'),
  pathToAppData: app.getPath('appData'),
  pathToAppDir: app.getAppPath(),
  pathToDocDir: app.getPath('documents'),
  pathToDwnDir: app.getPath('downloads'),
  pathToResources: process.resourcesPath,
  pathToTempDir: app.getPath('temp'),
  pathToUsrHome: app.getPath('home'),
  updateFeedUrl: autoUpdater.getFeedURL(),
  userLocale: app.getLocale(),
};

export default envProps;
