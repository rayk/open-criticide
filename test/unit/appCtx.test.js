import appCtx from '../../src/context/appCtx';
import { dev, prod } from '../stubs/envProps.stub';

const config = require('../../src/context/config.json');
const R = require('ramda');
const Po = require('ramda-adjunct');

describe('Application configuration values are set in a JSON file:', () => {
  it('structured as an Object.', () => {
    expect(config).not.toBeUndefined();
  });

  it('holds configurations values as integers.', () => {
    expect(Po.isInteger(config.windowDefaultHeight)).toBeTruthy();
  });

  it('holds configurations values as booleans.', () => {
    expect(Po.isBoolean(config.experimentalFeatures)).toBeTruthy();
  });
});

describe('Application Context (Test Hookup Exercise):', () => {
  it('has objects for App Details, Paths, Platform & Logger.', () => {
    const ctxKeys = R.keys(appCtx(dev, config));
    expect(
      R.equals(ctxKeys, [
        'app',
        'path',
        'platform',
        'logger',
        'isRunningOnMac',
        'isRunningOnWin',
        'isRunningOnLinux',
      ]),
    ).toBeTruthy();
    expect(R.values(appCtx(dev, config))).not.toContain(undefined);
  });

  it('has attributes for App Details in the following order.', () => {
    const ctxApp = appCtx(dev, config).app;
    const appKeys = R.keys(ctxApp) /* ? */;
    expect(
      R.equals(appKeys, [
        'name',
        'updateUrl',
        'productName',
        'version',
        'description',
        'experimentalFeatures',
        'released',
        'rollBarMain',
        'rollBarRenderer',
        'scrollBounce',
        'windowDefaultHeight',
        'windowDefaultWidth',
        'windowMainContent',
        'windowMinHeight',
        'windowMinWidth',
        'windowTitle',
      ]),
    ).toEqual(true);
    expect(R.values(ctxApp)).not.toContain(undefined);
  });

  it('has attributes for App Directory Paths.', () => {
    const ctxPath = appCtx(dev, config).path;
    const pathKeys = R.keys(ctxPath);
    expect(
      R.equals(pathKeys, [
        'appFiles',
        'configData',
        'dataFiles',
        'documents',
        'downloads',
        'home',
        'resources',
        'tempData',
      ]),
    ).toBeTruthy();
    expect(R.values(ctxPath)).not.toContain(undefined);
  });

  it('has attributes for App Platform Details.', () => {
    const ctxPlatform = appCtx(dev, config).platform;
    const platformKeys = R.keys(ctxPlatform);
    expect(
      R.equals(platformKeys, [
        'macStoreBuild',
        'winStoreBuild',
        'chromeVersion',
        'electronVersion',
        'os',
        'processor',
        'devMode',
      ]),
    ).toBeTruthy();
    expect(R.values(ctxPlatform)).not.toContain(undefined);
  });

  it('has configured logger object.', () => {
    const ctxDev = appCtx(dev, config);
    const ctxProd = appCtx(prod, config);
    const loggerDev = ctxDev.logger;
    const loggerPrd = ctxProd.logger;
    expect(loggerDev).toBeDefined();
    expect(loggerPrd).toBeDefined();
    if (ctxDev.platform.devMode) {
      expect(loggerDev.transports.file.level).toEqual('silly');
      expect(loggerDev.transports.console.level).toEqual('silly');
    } else {
      expect(loggerPrd.transports.file.level).toEqual('warn');
      expect(loggerPrd.transports.file.console).toEqual('warn');
    }
  });
});

describe('Application Context is not obscured:', () => {
  it('development context generally matches this snapshot', () => {
    expect(appCtx(dev, config)).toMatchSnapshot();
  });
  it('production context generally matches this snapshot', () => {
    expect(appCtx(prod, config)).toMatchSnapshot();
  });
});
