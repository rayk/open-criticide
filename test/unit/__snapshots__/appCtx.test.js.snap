// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Application Context is not obscured: development context generally matches this snapshot 1`] = `
Object {
  "app": Object {
    "description": "Criticide Probability Engine Desktop Application.",
    "experimentalFeatures": true,
    "name": "criticide",
    "productName": "Criticide",
    "released": "Mon Jun 05 2017 14:19:31 GMT+1000 (AEST)",
    "rollBarMain": "b6cedc15e5b94d88a74cd6f0cbf579e5",
    "rollBarRenderer": "2509240345f14c718038f8a7018c11a9",
    "scrollBounce": true,
    "updateUrl": "www.criticide.com/some/path/to/feed/url",
    "version": "0.0.1",
    "windowDefaultHeight": 800,
    "windowDefaultWidth": 1024,
    "windowMainContent": "startup/index.html",
    "windowMinHeight": 600,
    "windowMinWidth": 1024,
    "windowTitle": "Criticide",
  },
  "isRunningOnLinux": true,
  "isRunningOnMac": false,
  "isRunningOnWin": false,
  "logger": Object {
    "debug": [Function],
    "default": [Circular],
    "error": [Function],
    "info": [Function],
    "log": [Function],
    "silly": [Function],
    "transports": Object {
      "console": [Function],
      "file": [Function],
      "logS": [Function],
      "rendererConsole": [Function],
    },
    "verbose": [Function],
    "warn": [Function],
  },
  "path": Object {
    "appFiles": "/usr/local/userData",
    "configData": "/usr/local/appData",
    "dataFiles": "/usr/applications",
    "documents": "/usr/local/documents",
    "downloads": "/usr/local/downloads",
    "home": "usr/resources",
    "resources": "usr/temp",
    "tempData": "/usr",
  },
  "platform": Object {
    "chromeVersion": "56.0.2924.87",
    "devMode": true,
    "electronVersion": "1.6.10",
    "macStoreBuild": false,
    "os": "linux",
    "processor": "x64",
    "winStoreBuild": false,
  },
}
`;

exports[`Application Context is not obscured: production context generally matches this snapshot 1`] = `
Object {
  "app": Object {
    "description": "Criticide Probability Engine Desktop Application.",
    "experimentalFeatures": true,
    "name": "criticide",
    "productName": "Criticide",
    "released": "Mon Jun 05 2017 14:19:31 GMT+1000 (AEST)",
    "rollBarMain": "b6cedc15e5b94d88a74cd6f0cbf579e5",
    "rollBarRenderer": "2509240345f14c718038f8a7018c11a9",
    "scrollBounce": true,
    "updateUrl": "www.criticide.com/some/path/to/feed/url",
    "version": "0.0.1",
    "windowDefaultHeight": 800,
    "windowDefaultWidth": 1024,
    "windowMainContent": "startup/index.html",
    "windowMinHeight": 600,
    "windowMinWidth": 1024,
    "windowTitle": "Criticide",
  },
  "isRunningOnLinux": true,
  "isRunningOnMac": false,
  "isRunningOnWin": false,
  "logger": Object {
    "debug": [Function],
    "default": [Circular],
    "error": [Function],
    "info": [Function],
    "log": [Function],
    "silly": [Function],
    "transports": Object {
      "console": [Function],
      "file": [Function],
      "logS": [Function],
      "rendererConsole": [Function],
    },
    "verbose": [Function],
    "warn": [Function],
  },
  "path": Object {
    "appFiles": "/usr/local/userData",
    "configData": "/usr/local/appData",
    "dataFiles": "/usr/applications",
    "documents": "/usr/local/documents",
    "downloads": "/usr/local/downloads",
    "home": "usr/resources",
    "resources": "usr/temp",
    "tempData": "/usr",
  },
  "platform": Object {
    "chromeVersion": "56.0.2924.87",
    "devMode": false,
    "electronVersion": "1.6.10",
    "macStoreBuild": false,
    "os": "linux",
    "processor": "x64",
    "winStoreBuild": false,
  },
}
`;
